[pubsub] add output plugin

dmichel1 opened the pull request:
This plugin allows for outputting metrics to Google Pubsub.

It allows for local testing via the pubsub emulator. See the readme on setup instructions. 


 
Threading appears to be setup correctly as not to block the main ffwd thread (looks okay) but the grpc thread count doesn't seem to be tunable (?). Let's see how this runs on a few instances internally.

<img width="1081" alt="screen shot 2018-07-20 at 1 37 41 pm" src="https://user-images.githubusercontent.com/677960/43018152-ac49d538-8c26-11e8-9dc7-27d537c704ee.png">

dmichel1 committed:
[pubsub] add output plugin

This plugin allows for outputting metrics to Google Pubsub.

dmichel1 committed:
[riemann] name module correctly

jsferrei commented on modules/pubsub/src/main/java/com/spotify/ffwd/pubsub/PubsubPluginSink.java:
```diff
@@ -0,0 +1,142 @@
+    final ArrayList<Metric> metrics = new ArrayList<>();
+
+    batches.forEach(batch -> batch.getPoints().forEach(
+        point -> metrics.add(convertBatchMetric(batch, point))));
```
Since this is a nested `forEach`, I think it would be more readable  formatted as
```
batch -> {
   batch.getPoints().forEach(point -> {
      metrics.add(convertBatchMetric(batch, point));
   }
}
```

dmichel1 replied:
fixed formatting 👍

jsferrei commented on modules/pubsub/src/main/java/com/spotify/ffwd/pubsub/PubsubOutputPlugin.java:
```diff
@@ -0,0 +1,147 @@
+        Preconditions.checkArgument(topic.isPresent(), "Pubsub topic must be set in config");
+
+        final Duration publishDelayThreshold = Duration.ofNanos(
+            publishDelayThresholdMs.orElse(DEFAULT_DELAY_THRESHOLD));
```
Perhaps push this `orElse` to the constructor as you did with `this.messageCountBatchSize`

dmichel1 replied:
Fixed.

jsferrei reviewed (approved):
I think this looks fine, just had a couple of comments

dmichel1 commented:
@jsferrei thanks - I just need to add tests and figure out the async part and then I think this is ready to test on a few instances.

sjoeboo commented:
So i was just poking around with this and 2 things: 

1). looks like this won't read the env vars to send to the local emulator if set: https://cloud.google.com/pubsub/docs/emulator#accessing_environment_variables

2). since ^^ wasn't working, FFWD was sending metrics basically no where because the test project i had set for the emulator doesn't exist in the "real" PubSub, yet FFWD threw no errors about it not being able to send metrics anywhere.

dmichel1 commented:
@sjoeboo I added emulator support (also in the readme on how to setup) and added a warning on startup if the topic doesn't exist. 

@hexedpackets was working on having the heroic consumers create the topics if they don't exist. The reason the publishes will not is because they will all be running with the least amount of privileges. Also longer term we were thinking about managing the topics and subscriptions in terraform.

sjoeboo commented:
+1
